### ISSUE TITLE
Groovy Filtering issues (and Dynamic configuration)

### ISSUE THREAD

**Issue by ildar2** [MarathonLabs/marathon#752]
Test filtering isn't working as expected when using gradle plugin with groovy scripts

I'm trying to configure filteringConfiguration in my groovy scripts using Wrapper class' parameters:

```
  filteringConfiguration {
    allowlist {
      annotationFilter = findProperty('annotationFilter')?.split(',') ?: ['androidx.test.filters.MediumTest']// only default works, set from property it launches all tests
      testMethodFilter = findProperty('testMethodFilter')?.tokenize(',') ?: ['failedRestaurantOrder']//doesn't work in both variants - Scheduling 0 tests
      fullyQualifiedClassnameFilter = ['com.package.AgreementEspressoTest', 'com.package.PromosOnSummaryEspressoTest']//doesn't work - Scheduling 0 tests
      simpleClassNameFilter = ['AgreementEspressoTest', 'PromosOnSummaryEspressoTestKt']//doesn't work - Scheduling 0 tests
    }
  }
```
some of them seem to not work at all (simpleClassNameFilter, fullyQualifiedClassnameFilter, testMethodFilter), some are missing (fully-qualified-test-name), none working if set from property as a list - I've tried split(',') and tokenize(',') to no avail

What I did:
```
task1:
'marathonDebugAndroidTest -PannotationFilter=androidx.test.filters.MediumTest'
config:
    allowlist {
      annotationFilter = findProperty('annotationFilter')?.split(',') ?: ['androidx.test.filters.LargeTest']
      println(annotationFilter)//[androidx.test.filters.MediumTest]
    }
result: 
I 14:07:49.615 [main @coroutine#1] <com.malinskiy.marathon.Marathon> Scheduling 487 tests (all tests - incorrect)
```

```
task2:
'marathonDebugAndroidTest -PtestMethodFilter=failedRestaurantOrder,checkPromoBlockDisabledExp
config:
    allowlist {
      testMethodFilter = findProperty('testMethodFilter')?.tokenize(',') ?: []
      println(testMethodFilter)//[failedRestaurantOrder, checkPromoBlockDisabledExp]

//      testMethodFilter = ['failedRestaurantOrder', 'checkPromoBlockDisabledExp']//also doesn't work
    }
result: 
I 13:51:44.318 [main @coroutine#1] <com.malinskiy.marathon.Marathon> Scheduling 0 tests (incorrect, should be 2)
```

also I've tried fully qualified test names, tried to add Kt to class names, but nothing worked

**Comment by Malinskiy** [comment]
@ildar2 the problem you're having has nothing to do with marathon, it's a configuration of gradle that you need to figure out.

I can only suggest looking at the generated Marathonfile in the build directory to verify what is the actual filteringConfiguration you're generating.

The snippets that you provide do not show anything about the values you're passing into the properties, it is also complicated by the fact that all the filters you're specifying in groovy are just `ArrayList<String>`, so they're not representative of the final filtering configuration, that's why looking on the generated Marathonfile is the best way forward.

I can take a look at your gradle config code if you can publish the gradle project somewhere

**Comment by ildar2** [comment]
Thanks for the tip with the generated Marathonfile! 

Well, some of it has to do with marathon, because config maps incorrectly to Marathonfile

my config: 
```

  filteringConfiguration {
    allowlist {
      annotationFilter = findProperty('annotationFilter')?.split(',') ?: ['androidx.test.filters.LargeTest']
      println(annotationFilter)//[androidx.test.filters.MediumTest]
      testMethodFilter = findProperty('testMethodFilter')?.tokenize(',') ?: ['com.package.QRPaymentEspressoTest#failedRestaurantOrder', 'PromosOnSummaryEspressoTest#checkPromoBlockDisabledExp']
      println(testMethodFilter)//[failedRestaurantOrder, checkPromoBlockDisabledExp]
      fullyQualifiedClassnameFilter = ['com.package.AgreementEspressoTest', 'com.package.PromosOnSummaryEspressoTest']
      println(fullyQualifiedClassnameFilter)//[com.package.AgreementEspressoTest, com.package.PromosOnSummaryEspressoTest]
      simpleClassNameFilter = ['AgreementEspressoTest', 'PromosOnSummaryEspressoTest']
      println(simpleClassNameFilter)//[AgreementEspressoTest, PromosOnSummaryEspressoTest]
    }
  }
```

for task:
`marathonDebugAndroidTest -PannotationFilter=androidx.test.filters.MediumTest -PtestMethodFilter=failedRestaurantOrder,checkPromoBlockDisabledExp`

resulted in this:
```
filteringConfiguration:
  allowlist:
  - type: "fully-qualified-class-name"
    regex: "com.package.AgreementEspressoTest"
  - type: "fully-qualified-class-name"
    regex: "com.package.PromosOnSummaryEspressoTest"
  - type: "simple-class-name"
    regex: "AgreementEspressoTest"
  - type: "simple-class-name"
    regex: "PromosOnSummaryEspressoTestKt"
```

`fully-qualified-class-name` and `simple-class-name` mapped as two entries with `regex` instead of `values`. And it lost `annotationFilter` and `testMethodFilter`. `println(annotationFilter)` clearly printed out `[androidx.test.filters.MediumTest]`, but it didn't show up in Marathonfile - I can't figure it out, might be different array/list classes. Same with testMethodFilter

The problem is with this file: https://github.com/MarathonLabs/marathon/blob/develop/marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/gradle/configuration/FilteringPluginConfiguration.kt
it doesn't have `fully-qualified-test-name` and the `Wrapper` class is mapped incorrectly to Kotlin classes

**Comment by Malinskiy** [comment]
The gradle plugin config for filtering hasn't been changed for a long time. I suggest trying to call the kts methods https://github.com/MarathonLabs/marathon/blob/develop/marathon-gradle-plugin/src/main/kotlin/com/malinskiy/marathon/gradle/configuration/FilteringPluginConfiguration.kt#L25 somehow from groovy and adding a specific instance of filter there. If that's not possible then your only option is changing the code in marathon. I'd be happy to review a PR with a fix for the groovy config if you have the time and energy to submit it.

**Comment by ildar2** [comment]
@Malinskiy I've submitted a PR that would fix multiple values per filtering configuration.

Also I've discovered why something like `annotationFilter = findProperty('annotationFilter')?.split(',') ?: []` wouldn't properly set that value. findProperty('annotationFilter') for some reason would pull `filteringConfiguration.groovyAllowList.annotationFilter` itself instead of a param submitted by `-PannotationFilter=androidx.test.filters.MediumTest` resulting in it being set from null to itself (null) again. That's weird, although it kinda says about that in gradle docs.
The logical solution is to rename param to something like  `-PmarathonAnnotations=androidx.test.filters.MediumTest ` so `annotationFilter = findProperty('marathonAnnotations')?.split(',') ?: []` would properly set `annotationFilter` value.
